src/tensors/treetransformers.jl

```diff
@@ -7,68 +7,136 @@ abstract type TreeTransformer end
-    F3 = fusiontreetype(I, numout(Vsrc))
-    F4 = fusiontreetype(I, numin(Vsrc))
-
-    if FusionStyle(I) isa UniqueFusion
-        return AbelianTreeTransformer{sectorscalartype(I),I,N,F1,F2,F3,F4}
-    else
-        return GenericTreeTransformer{sectorscalartype(I),I,N,F1,F2,F3,F4}
+    for i in 1:L
+        f₁, f₂ = structure_src.fusiontreelist[i]
+        (f₃, f₄), coeff = only(transform(f₁, f₂))
+        j = structure_dst.fusiontreeindices[(f₃, f₄)]
+        stridestructure_dst = structure_dst.fusiontreestructure[j]
+        stridestructure_src = structure_src.fusiontreestructure[i]
+        data[i] = (coeff, stridestructure_dst, stridestructure_src)
     end
+
+    return AbelianTreeTransformer(data)
 end
 
-function TreeTransformer(transform::Function, Vdst::HomSpace{S},
-                         Vsrc::HomSpace{S}) where {S}
-    I = sectortype(Vdst)
-    I === Trivial && return TrivialTreeTransformer()
+struct GenericTreeTransformer{T,N} <: TreeTransformer
+    data::Vector{Tuple{Matrix{T},Vector{StridedStructure{N}},Vector{StridedStructure{N}}}}
+end
 
+function GenericTreeTransformer(transform, p, Vdst, Vsrc)
+    permute(Vsrc, p) == Vdst || throw(SpaceMismatch("Incompatible spaces for permuting."))
     structure_dst = fusionblockstructure(Vdst)
     structure_src = fusionblockstructure(Vsrc)
+    I = sectortype(Vsrc)
+
+    uncoupleds_src = map(structure_src.fusiontreelist) do (f₁, f₂)
+        return TupleTools.vcat(f₁.uncoupled, f₂.uncoupled)
+    end
+    uncoupleds_src_unique = unique(uncoupleds_src)
 
-    rows = Int[]
-    cols = Int[]
-    vals = sectorscalartype(sectortype(Vdst))[]
+    uncoupleds_dst = map(structure_dst.fusiontreelist) do (f₁, f₂)
+        return TupleTools.vcat(f₁.uncoupled, f₂.uncoupled)
+    end
+
+    T = sectorscalartype(I)
+    N = numind(Vdst)
+    L = length(uncoupleds_src_unique)
+    TStrided = StridedStructure{N}
+    data = Vector{Tuple{Matrix{T},Vector{TStrided},Vector{TStrided}}}(undef, L)
 
-    for (row, (f1, f2)) in enumerate(structure_src.fusiontreelist)
-        for ((f3, f4), coeff) in transform(f1, f2)
-            col = structure_dst.fusiontreeindices[(f3, f4)]
-            push!(rows, row)
-            push!(cols, col)
-            push!(vals, coeff)
+    # TODO: this can be multithreaded
+    for (i, uncoupled) in enumerate(uncoupleds_src_unique)
+        ids_src = findall(==(uncoupled), uncoupleds_src)
+        fusiontrees_outer_src = structure_src.fusiontreelist[ids_src]
+
+        uncoupled_dst = TupleTools.getindices(uncoupled, (p[1]..., p[2]...))
+        ids_dst = findall(==(uncoupled_dst), uncoupleds_dst)
+
+        fusiontrees_outer_dst = structure_dst.fusiontreelist[ids_dst]
+
+        matrix = zeros(sectorscalartype(I), length(ids_dst), length(ids_src))
+        for (row, (f₁, f₂)) in enumerate(fusiontrees_outer_src)
+            for ((f₃, f₄), coeff) in transform(f₁, f₂)
+                col = findfirst(==((f₃, f₄)), fusiontrees_outer_dst)::Int
```

Comment:
Rather than asserting `::Int`, should we add an error check?
```julia
if isnothing(col)
    error("unexpected...")
end
```
Then, the Julia compiler should know that `col` is an `Int` beyond this point.

Reply:
Although I guess the type assert accomplishes the same thing; it will error if `findfirst` returns `nothing`, and since this shouldn't happen anyway, it doesn't matter if the error is a bit cryptic.

Reply:
Yes, this was the idea of that type assert. I don't think it is required to provide a better error either, so I'll keep the current implementation.